input-profile.js

```diff
@@ -0,0 +1,249 @@
+import {Component, Property} from '@wonderlandengine/api';
+import {vec3, quat} from 'gl-matrix';
+const _TempVec = vec3.create();
+const _TempQuat = quat.create();
+
+export class InputProfile extends Component {
+    static TypeName = 'input-profile';
+    static Properties = {
+        controller: Property.object(),
+        handednessIndex: Property.enum(['left', 'right'], 0),
+        path: Property.string(
+            'https://cdn.jsdelivr.net/npm/@webxr-input-profiles/assets@1.0.13/dist/profiles/'
+        ),
+        customProfileFolder: Property.string(''),
+        preSpawnedController: Property.object(),
+    };
+
+    init() {
+        this._gamepadObjects = {};
+        this._modelLoaded = false;
+        this.controllerDeactivateFlag = false;
+        this.ProfileJSON = {};
+    }
+
+    start() {
+        if (this.engine.XRSession) {
+            this.engine.XRSession.addEventListener(
+                'inputsourceschange',
+                this.onInputSourcesChange.bind(this)
+            );
+        } else {
+            this.engine.onXRSessionStart.add((session) => {
+                session.addEventListener(
+                    'inputsourceschange',
+                    this.onInputSourcesChange.bind(this)
+                );
+            });
+        }
+
+        this.handedness = ['left', 'right'][this.handednessIndex];
+
+        this.components = [];
+        if (this.preSpawnedController != null) {
+            this.getComponents(this.preSpawnedController);
+        }
+    }
+
+    update() {
+        if (this.modelLoaded) {
+            this.mapGamepadInput();
+        }
+    }
+
+    onInputSourcesChange(event) {
+        event.added.forEach((xrInputSource) => {
+            if (xrInputSource.hand != undefined) {
+                this.handleHandSource();
+                return;
+            } else {
+                this.handleControllerSource(xrInputSource);
+            }
+        });
+    }
+
+    handleControllerSource(xrInputSource) {
+        this.activatePrespawnedControllersOnStart();
+        this.loadProfileJson(xrInputSource);
+    }
+
+    handleHandSource(xrInputSource) {
+        this.deactivatePreSpawnedControllerOnStart();
+    }
+
+    activatePrespawnedControllersOnStart() {
+        this.setComponentsActive(true);
+    }
+
+    deactivatePreSpawnedControllerOnStart() {
+        this.setComponentsActive(false);
+    }
+
+    getComponents(obj) {
+        const comps = obj.getComponents().filter((c) => c.type !== 'vr-mode-active-switch');
```

Comment:
Hummm, what happened if someone has another component that shouldn't be activated either?

Reply:
hmm I think its not practical to provide such insurance.
Is there any suggestions ?

Reply:
And how can we give the user the control to decide that ? String ? Ain't it gonna further complex things up

Reply:
Just an array, not everything needs to be a property exposed to the user. It's a public attribute that is only accessible programmatically

Reply:
done